### PR TITLE
Validate mail address and mail domain

### LINKED ISSUE #28436: Admin Settings: Email From address takes any value
In the admin general settings, section email server _from address_ is accepting any character as valid value. Even when I enter an @ into both fields this is accepted.

This issue stumbled upon me, when a client entered a trailing comma for the user part by mistake (which prevented the mailer from sending emails)

There should be a check on valid email addresses.

### PR DESCRIPTION
**Pull request by imujjwal96**
## Description

Using regular expressions to validate  `$mail_from_address` and `$mail_domain`
## Related Issue

Fixes #28436 

## How Has This Been Tested?

Tested regex here: http://www.phpliveregex.com/

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by phil-davis** [comment]
This PR also got a Jenkins problem with ``sharing-v1.feature:914``
There seems to be a lot of this now???

File: settings/Controller/MailSettingsController.php [hunk 1]
```diff
@@ -104,6 +105,27 @@ public function setMailSettings($mail_domain,
 			$configs[$key] = (empty($value)) ? null : $value;
 		}
 
+		$regex_address = "/(?:[a-z0-9!#$%&'*+\/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+\/=?^_`{|}~-]+)*|\"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*\")/";
+		if (!preg_match($regex_address, $mail_from_address, $matches)) {
+			return ['data' =>
+				['message' =>
+					(string) $this->l10n->t('Invalid Address')
```
Comment (mmattel) [inline settings/Controller/MailSettingsController.php]: Text proposal: ``Invalid Address`` --> ``Invalid eMail recipient``

File: settings/Controller/MailSettingsController.php [hunk 2]
```diff
@@ -104,6 +105,27 @@ public function setMailSettings($mail_domain,
...
+				],
+				'status' => 'error'
+			];
+		}
+
+		$regex_domain = "/(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])/";
+		if (!preg_match($regex_domain, $mail_domain, $matches)) {
+			return ['data' =>
+				['message' =>
+					(string) $this->l10n->t('Invalid Domain')
```
Comment (mmattel) [inline settings/Controller/MailSettingsController.php]: Text proposal: ``Invalid Domain`` --> ``Invalid eMail domain``

File: settings/Controller/MailSettingsController.php [hunk 3]
```diff
@@ -104,6 +105,27 @@ public function setMailSettings($mail_domain,
 			$configs[$key] = (empty($value)) ? null : $value;
 		}
 
+		$regex_address = "/(?:[a-z0-9!#$%&'*+\/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+\/=?^_`{|}~-]+)*|\"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*\")/";
```
Comment (DeepDiver1975) [inline settings/Controller/MailSettingsController.php]: Please use IMailer::validateMailAddress function for this - THX

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

File: settings/Controller/MailSettingsController.php (hunk 3 above)
Comment (imujjwal96) [reply settings/Controller/MailSettingsController.php]: We are taking two separate inputs for `$mail_from_address` and `$mail_domain`. Should I merge them into a single input in the UI?

Comment (imujjwal96) [reply settings/Controller/MailSettingsController.php]: @DeepDiver1975 I've updated it. Though i didn't make any change in the UI yet.

**Comment by PVince81** [comment]
```
Tests\Settings\Controller\MailSettingsControllerTest::testSetMailSettings
Failed asserting that Array &0 (
    'data' => Array &1 (
        'message' => 'Saved'
    )
    'status' => 'error'
) is identical to Array &0 (
    'data' => Array &1 (
        'message' => 'Saved'
    )
    'status' => 'success'
).

/var/lib/jenkins/workspace/owncloud-core_core_PR-28948-II5ZCIFKAAAJDW3PM64ZM7S65WOXKGXVTLCQ5NPVX6GY3POENMVA/tests/Settings/Controller/MailSettingsControllerTest.php:120
```

@imujjwal96 can you have a look ? Also once fixed, please squash your commits into a single one. Thanks

**Comment by imujjwal96** [comment]
@PVince81 Is this the expected value? 
image

The test case expects it as
`demo@owncloud.com `  @ `owncloud.com`
image

**Comment by PVince81** [comment]
@imujjwal96 hmm maybe that old test was wrong

**Comment by PVince81** [comment]
kicked Jenkins

**Comment by PVince81** [comment]
Taking over for rebasing, here: https://github.com/owncloud/core/pull/29909

**Pull request closed without merging**